### PR TITLE
Refactor parameters for bootstrapping

### PR DESCRIPTION
**Pull request by crockeea**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by crockeea** [comment]
CI will fail until the [latticpp PR](https://github.com/awslabs/aws-cppwrapper-lattigo/pull/6) is merged.

File: src/hit/api/context.h [hunk 1]
```diff
@@ -54,6 +55,8 @@ namespace hit {
         int log_scale() const;
 
         latticpp::Parameters params;
+        std::optional<latticpp::BootstrappingParameters> btp_params =
+            std::optional<latticpp::BootstrappingParameters>();
```
[the remark this answers is not in the document]
Comment (crockeea) [reply src/hit/api/context.h:59]: I don't think so. Removed.

File: src/hit/api/params.cpp [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    //       maxN1N2Ratio(maxN1N2Ratio) {
+    //     // From https://eprint.iacr.org/eprint-bin/getfile.pl?entry=2020/1203&version=20201008:204443&file=1203.pdf
+    //     // section 5.4 The latest version of the paper https://eprint.iacr.org/2020/1203.pdf doesn't go into as much
+    //     // detail and merely says "use Cos2 for non-sparse keys" (section 5.4)
+    //     if (approx_degree < 2 * approx_range - 1) {
+    //         trig_approx = Cos2;
+    //     } else {
+    //         trig_approx = Cos1;
+    //     }
+    // }
```
[the remark this answers is not in the document]
Comment (crockeea) [reply src/hit/api/params.cpp:29]: Nope, just an outdated comment. Removed.

**Review by SalusaSecondus — APPROVED** [review]

File: src/hit/api/context.cpp [hunk 3]
```diff
@@ -83,15 +68,10 @@ namespace hit {
...
-    HEContext::HEContext(Parameters &params) : params(move(params)) {
-        validateContext();
-    }
-
-    HEContext::HEContext(int num_slots, int mult_depth, int precisionBits) {
-        vector<uint8_t> logQi = gen_ciphertext_modulus_vec(mult_depth + 1, precisionBits);
-        vector<uint8_t> logPi(1);
-        logPi[0] = 60;  // special modulus. For now, we just use a single modulus like SEAL.
-        params = newParametersFromLogModuli(log2(num_slots) + 1, logQi, mult_depth + 1, logPi, 1, precisionBits);
+    HEContext::HEContext(const CKKSParams &params) : params(move(params.lattigo_params)) {
```
[the remark this answers is not in the document]
Comment (crockeea) [reply src/hit/api/context.cpp:71]: Good catch. Will fix.

File: src/hit/api/params.cpp [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    CKKSParams::CKKSParams(latticpp::Parameters params) : lattigo_params(move(params)) {
+    }
+
+    CKKSParams::CKKSParams(latticpp::Parameters lattigo_params, latticpp::BootstrappingParameters lattigo_btp_params)
+        : lattigo_params(move(lattigo_params)),
+          btp_params(optional<BootstrappingParams>(BootstrappingParams(move(lattigo_btp_params)))) {
+    }
+
+    CKKSParams::CKKSParams(int num_slots, int log_scale, int max_ct_level, int num_ks_primes,
```
[the remark this answers is not in the document]
Comment (crockeea) [reply src/hit/api/params.cpp:25]: `log_scale` will always be in the range ~[20,60], because backends limit the size of an individual modulus to 64 bits.

**Pull request merged**